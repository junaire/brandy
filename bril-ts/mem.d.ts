/**
 * https://capra.cs.cornell.edu/bril/lang/memory.html
 */
export interface Pointer<T> {}
export function alloc<T>(size: bigint): Pointer<T>
export function store<T>(pointer: Pointer<T>, value: T)
export function load<T>(pointer: Pointer<T>): T
export function free<T>(pointer: Pointer<T>): void;
export function ptradd<T>(pointer: Pointer<T>, offset: bigint): Pointer<T>
