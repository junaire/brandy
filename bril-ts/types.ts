import * as bril from './bril.ts';

/**
 * An abstract type signature.
 *
 * Describes the shape and types of all the ingredients for a Bril operation
 * instruction: arguments, result, labels, and functions.
 */
export interface BaseSignature<T> {
  /**
   * The types of each argument to the operation.
   */
  args: T[],

  /**
   * The result type, if non-void.
   */
  dest?: T,

  /**
   * The number of labels required for the operation.
   */
  labels?: number,

  /**
   * The number of function names required for the operation.
   */
  funcs?: number,
}

/**
 * The concrete type signature for an operation.
 */
export type Signature = BaseSignature<bril.Type>;

/**
 * A polymorphic type variable.
 */
export type TVar = {"tv": string};

/**
 * Like bril.Type, except that type variables may occur at the leaves.
 */
export type PolyType = bril.PrimType | TVar | {"ptr": PolyType};

/**
 * A polymorphic type signature, universally quantified over a single
 * type variable.
 */
export interface PolySignature {
    tvar: TVar;
    sig: BaseSignature<PolyType>;
}

/**
 * The type of a Bril function.
 */
export interface FuncType {
  /**
   * The types of the function's arguments.
   */
  args: bril.Type[],

  /**
   * The function's return type.
   */
  ret: bril.Type | undefined,
}

/**
 * Type signatures for the Bril operations we know.
 */
export const OP_SIGS: {[key: string]: Signature | PolySignature} = {
  // Core.
  'add': {args: ['int', 'int'], dest: 'int'},
  'mul': {args: ['int', 'int'], dest: 'int'},
  'sub': {args: ['int', 'int'], dest: 'int'},
  'div': {args: ['int', 'int'], dest: 'int'},
  'eq': {args: ['int', 'int'], dest: 'bool'},
  'lt': {args: ['int', 'int'], dest: 'bool'},
  'gt': {args: ['int', 'int'], dest: 'bool'},
  'le': {args: ['int', 'int'], dest: 'bool'},
  'ge': {args: ['int', 'int'], dest: 'bool'},
  'not': {args: ['bool'], dest: 'bool'},
  'and': {args: ['bool', 'bool'], dest: 'bool'},
  'or': {args: ['bool', 'bool'], dest: 'bool'},
  'jmp': {args: [], 'labels': 1},
  'br': {args: ['bool'], 'labels': 2},
  'id': {tvar: {tv: 'T'}, sig: {args: [{tv: 'T'}], dest: {tv: 'T'}}},
  'nop': {args: []},

  // Floating point.
  'fadd': {args: ['float', 'float'], dest: 'float'},
  'fmul': {args: ['float', 'float'], dest: 'float'},
  'fsub': {args: ['float', 'float'], dest: 'float'},
  'fdiv': {args: ['float', 'float'], dest: 'float'},
  'feq': {args: ['float', 'float'], dest: 'bool'},
  'flt': {args: ['float', 'float'], dest: 'bool'},
  'fgt': {args: ['float', 'float'], dest: 'bool'},
  'fle': {args: ['float', 'float'], dest: 'bool'},
  'fge': {args: ['float', 'float'], dest: 'bool'},

  // Memory.
  'alloc': {tvar: {tv: 'T'}, sig: {args: ['int'], dest: {ptr: {tv: 'T'}}}},
  'free': {tvar: {tv: 'T'}, sig: {args: [{ptr: {tv: 'T'}}]}},
  'store': {tvar: {tv: 'T'}, sig: {args: [{ptr: {tv: 'T'}}, {tv: 'T'}]}},
  'load': {tvar: {tv: 'T'}, sig: {args: [{ptr: {tv: 'T'}}], dest: {tv: 'T'}}},
  'ptradd': {tvar: {tv: 'T'}, sig: {args: [{ptr: {tv: 'T'}}, 'int'], dest: {ptr: {tv: 'T'}}}},

  // Speculation.
  'speculate': {args: []},
  'commit': {args: []},
  'guard': {args: ['bool'], labels: 1},

  // Character.
  'ceq': {args: ['char', 'char'], dest: 'bool'},
  'clt': {args: ['char', 'char'], dest: 'bool'},
  'cgt': {args: ['char', 'char'], dest: 'bool'},
  'cle': {args: ['char', 'char'], dest: 'bool'},
  'cge': {args: ['char', 'char'], dest: 'bool'},
  'char2int': {args: ['char'], dest: 'int'},
  'int2char': {args: ['int'], dest: 'char'},
};
