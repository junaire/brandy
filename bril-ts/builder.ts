import * as bril from './bril.ts';

/**
 * A utility for building up Bril programs.
 */
export class Builder {
  /**
   * The program we have built so far.
   */
  public program: bril.Program = { functions: [] };

  private curFunction: bril.Function | null = null;
  private nextFresh: number = 0;

  /**
   * Create a new, empty function into which further code will be generated.
   */
  buildFunction(name: string, args: bril.Argument[], type?: bril.Type) {
    let func: bril.Function;
    if (type === undefined) {
      func = {name: name, instrs: [], args: args};
    } else {
      func = {name: name, instrs: [], args: args, type: type};
    }
    this.program.functions.push(func);
    this.curFunction = func;
    this.nextFresh = 0;
    return func;
  }

  /**
   * Build an operation instruction that produces a result. If the name is
   * omitted, a fresh variable is chosen automatically.
   */
  buildValue(op: bril.ValueOpCode, type: bril.Type,
             args: string[], funcs?: string[], labels?: string[],
             dest?: string) {
    dest = dest || this.freshVar();
    let instr: bril.ValueOperation = { op, dest, type, args, funcs, labels };
    this.insert(instr);
    return instr;
  }

  /**
   * Build a non-value-producing (side-effecting) operation instruction.
   */
  buildEffect(op: bril.EffectOpCode,
              args: string[], funcs?: string[], labels?: string[]) {
    let instr: bril.EffectOperation = { op, args, funcs, labels };
    this.insert(instr);
    return instr;
  }

  /**
   * Build a function call operation. If a type is specified, the call
   * produces a return value.
   */
  buildCall(func: string, args: string[],
            type: bril.Type, dest?: string): bril.ValueOperation;
  buildCall(func: string, args: string[],
            type?: undefined, dest?: string): bril.EffectOperation;
  buildCall(func: string, args: string[],
            type?: bril.Type, dest?: string): bril.Operation {
    if (type) {
      return this.buildValue("call", type, args, [func], undefined, dest);
    } else {
      return this.buildEffect("call", args, [func], undefined);
    }
  }
  
  /**
   * Build a constant instruction. As above, the destination name is optional.
   */
  buildConst(value: bril.Value, type: bril.Type, dest?: string) {
    dest = dest || this.freshVar();
    let instr: bril.Constant = { op: "const", value, dest, type };
    this.insert(instr);
    return instr;
  }

  /**
   * Build a constant integer value.
   */
  buildInt(value: number, dest?: string) {
    return this.buildConst(value, "int", dest);
  }

  /**
   * Build a constant boolean value.
   */
  buildBool(value: boolean, dest?: string) {
    return this.buildConst(value, "bool", dest);
  }

  /**
   * Build a constant floating-point value.
   */
  buildFloat(value: number, dest?: string) {
    return this.buildConst(value, "float", dest);
  }

  /**
   * Build a constant character value.
   */
  buildChar(value: string, dest?: string) {
    return this.buildConst(value, "char", dest);
  }

  /**
   * Add a label to the function at the current position.
   */
  buildLabel(name: string) {
    let label = {label: name};
    this.insert(label);
  }

  /**
   * Insert an instruction at the end of the current function.
   */
  private insert(instr: bril.Instruction | bril.Label) {
    if (!this.curFunction) {
      throw "cannot build instruction/label without a function";
    }
    this.curFunction.instrs.push(instr);
  }

  /**
   * Generate an unused variable name.
   */
  freshVar() {
    let out = 'v' + this.nextFresh.toString();
    this.nextFresh += 1;
    return out;
  }

  /**
   * Generate an unused suffix.
   */
  freshSuffix() {
    let out = '.' + this.nextFresh.toString();
    this.nextFresh += 1;
    return out;
  }

  setCurrentFunction(func: bril.Function) {
    this.curFunction = func;
  }
}
