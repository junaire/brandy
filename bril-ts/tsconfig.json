{
    "compilerOptions": {
        "strict": true,
        "outDir": "build",
        "target": "es6",
        "module": "commonjs",
        "sourceMap": true
    }
}
