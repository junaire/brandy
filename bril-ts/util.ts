/**
 * Read all the data from stdin as a string.
 */
export async function readStdin(): Promise<string> {
  const buf = await Deno.readAll(Deno.stdin);
  return (new TextDecoder()).decode(buf);
}

export function unreachable(x: never): never {
  throw "impossible case reached";
}
