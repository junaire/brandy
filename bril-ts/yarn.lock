# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


"@types/node@^11.13.5":
  version "11.15.54"
  resolved "https://registry.yarnpkg.com/@types/node/-/node-11.15.54.tgz#59ed60e7b0d56905a654292e8d73275034eb6283"
  integrity sha512-1RWYiq+5UfozGsU6MwJyFX6BtktcT10XRjvcAQmskCtMcW3tPske88lM/nHv7BQG1w9KBXI1zPGuu5PnNCX14g==

typescript@^3.4.3:
  version "3.9.10"
  resolved "https://registry.yarnpkg.com/typescript/-/typescript-3.9.10.tgz#70f3910ac7a51ed6bef79da7800690b19bf778b8"
  integrity sha512-w6fIxVE/H1PkLKcCPsFqKE7Kv7QUwhU8qQY2MueZXWx5cPZdwFupLgKK3vntcK98BtNHZtAF4LA/yl2a7k8R6Q==
