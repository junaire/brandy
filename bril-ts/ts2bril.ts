#!/usr/bin/env node
import * as ts from 'typescript';
import * as bril from './bril';
import {Builder} from './builder';

const opTokens = new Map<ts.SyntaxKind, [bril.ValueOpCode, bril.Type]>([
  [ts.SyntaxKind.PlusToken,               ["add", "int"]],
  [ts.SyntaxKind.AsteriskToken,           ["mul", "int"]],
  [ts.SyntaxKind.MinusToken,              ["sub", "int"]],
  [ts.SyntaxKind.SlashToken,              ["div", "int"]],
  [ts.SyntaxKind.LessThanToken,           ["lt",  "bool"]],
  [ts.SyntaxKind.LessThanEqualsToken,     ["le",  "bool"]],
  [ts.SyntaxKind.GreaterThanToken,        ["gt",  "bool"]],
  [ts.SyntaxKind.GreaterThanEqualsToken,  ["ge",  "bool"]],
  [ts.SyntaxKind.EqualsEqualsToken,       ["eq",  "bool"]],
  [ts.SyntaxKind.EqualsEqualsEqualsToken, ["eq",  "bool"]],
]);

const opTokensFloat = new Map<ts.SyntaxKind, [bril.ValueOpCode, bril.Type]>([
  [ts.SyntaxKind.PlusToken,               ["fadd", "float"]],
  [ts.SyntaxKind.AsteriskToken,           ["fmul", "float"]],
  [ts.SyntaxKind.MinusToken,              ["fsub", "float"]],
  [ts.SyntaxKind.SlashToken,              ["fdiv", "float"]],
  [ts.SyntaxKind.LessThanToken,           ["flt",  "bool"]],
  [ts.SyntaxKind.LessThanEqualsToken,     ["fle",  "bool"]],
  [ts.SyntaxKind.GreaterThanToken,        ["fgt",  "bool"]],
  [ts.SyntaxKind.GreaterThanEqualsToken,  ["fge",  "bool"]],
  [ts.SyntaxKind.EqualsEqualsToken,       ["feq",  "bool"]],
  [ts.SyntaxKind.EqualsEqualsEqualsToken, ["feq",  "bool"]],
]);

function brilType(node: ts.Node, checker: ts.TypeChecker): bril.Type {
  let tsType = checker.getTypeAtLocation(node);
  if (tsType.flags & (ts.TypeFlags.Number | ts.TypeFlags.NumberLiteral)) {
    return "float";
  } else if (tsType.flags &
             (ts.TypeFlags.Boolean | ts.TypeFlags.BooleanLiteral)) {
    return "bool";
  } else if (tsType.flags &
             (ts.TypeFlags.BigInt | ts.TypeFlags.BigIntLiteral)) {
    return "int";
  } else {
    throw "unimplemented type " + checker.typeToString(tsType);
  }
}

/**
 * Compile a complete TypeScript AST to a Bril program.
 */
function emitBril(prog: ts.Node, checker: ts.TypeChecker): bril.Program {
  let builder = new Builder();
  builder.buildFunction("main", []);  // Main has no return type.

  function emitExpr(expr: ts.Expression): bril.ValueInstruction {
    switch (expr.kind) {
    case ts.SyntaxKind.NumericLiteral: {
      let lit = expr as ts.NumericLiteral;
      let val = parseFloat(lit.text);
      return builder.buildFloat(val);
    }

    case ts.SyntaxKind.BigIntLiteral: {
      let lit = expr as ts.BigIntLiteral;
      let val = parseInt(lit.text);
      return builder.buildInt(val);
    }

    case ts.SyntaxKind.TrueKeyword: {
      return builder.buildBool(true);
    }

    case ts.SyntaxKind.FalseKeyword: {
      return builder.buildBool(false);
    }

    case ts.SyntaxKind.Identifier: {
      let ident = expr as ts.Identifier;
      let type = brilType(ident, checker);
      return builder.buildValue("id", type, [ident.text]);
    }

    case ts.SyntaxKind.BinaryExpression:
      let bin = expr as ts.BinaryExpression;
      let kind = bin.operatorToken.kind;

      // Handle assignments.
      switch (kind) {
      case ts.SyntaxKind.EqualsToken:
        if (!ts.isIdentifier(bin.left)) {
          throw "assignment to non-variables unsupported";
        }
        let dest = bin.left as ts.Identifier;
        let rhs = emitExpr(bin.right);
        let type = brilType(dest, checker);
        return builder.buildValue("id", type, [rhs.dest], undefined, undefined, dest.text);
      }

      // Handle "normal" value operators.
      let op: bril.ValueOpCode;
      let type: bril.Type;
      if (brilType(bin.left, checker) === "float" ||
          brilType(bin.right, checker) === "float") {
        // Floating point operators.
        let p = opTokensFloat.get(kind);
        if (!p) {
          throw `unhandled FP binary operator kind ${kind}`;
        }
        [op, type] = p;
      } else {
        // Non-float.
        let p = opTokens.get(kind);
        if (!p) {
          throw `unhandled binary operator kind ${kind}`;
        }
        [op, type] = p;
      }

      let lhs = emitExpr(bin.left);
      let rhs = emitExpr(bin.right);
      return builder.buildValue(op, type, [lhs.dest, rhs.dest]);

    // Support call instructions---but only for printing, for now.
    case ts.SyntaxKind.CallExpression:
      let call = expr as ts.CallExpression;
      if (call.expression.getText() === "console.log") {
        let values = call.arguments.map(emitExpr);
        builder.buildEffect("print", values.map(v => v.dest));
        return builder.buildInt(0);  // Expressions must produce values.
      } else {
        // Recursively translate arguments.
        let values = call.arguments.map(emitExpr);

        // Check if effect statement, i.e., a call that is not a subexpression
        if (call.parent.kind === ts.SyntaxKind.ExpressionStatement) {
          builder.buildCall(call.expression.getText(), 
            values.map(v => v.dest));
          return builder.buildInt(0);  // Expressions must produce values
        } else {
          let decl = call.parent as ts.VariableDeclaration;
          let type = brilType(decl, checker);
          let name = (decl.name != undefined) ? decl.name.getText() : undefined;
          return builder.buildCall(
            call.expression.getText(), 
            values.map(v => v.dest), 
            type, 
            name,
          );
        } 
      }
    default:
      throw `unsupported expression kind: ${expr.getText()}`;
    }
  }

  function emit(node: ts.Node) {
    switch (node.kind) {
      // Descend through containers.
      case ts.SyntaxKind.SourceFile:
      case ts.SyntaxKind.Block:
      case ts.SyntaxKind.VariableStatement:
      case ts.SyntaxKind.VariableDeclarationList:
        ts.forEachChild(node, emit);
        break;

      // No-op.
      case ts.SyntaxKind.EndOfFileToken:
        break;

      // Emit declarations.
      case ts.SyntaxKind.VariableDeclaration: {
        let decl = node as ts.VariableDeclaration;
        // Declarations without initializers are no-ops.
        if (decl.initializer) {
          let init = emitExpr(decl.initializer);
          let type = brilType(decl, checker);
          builder.buildValue("id", type, [init.dest], undefined, undefined, decl.name.getText());
        }
        break;
      }

      // Expressions by themselves.
      case ts.SyntaxKind.ExpressionStatement: {
        let exstmt = node as ts.ExpressionStatement;
        emitExpr(exstmt.expression);  // Ignore the result.
        break;
      }

      // Conditionals.
      case ts.SyntaxKind.IfStatement: {
        let if_ = node as ts.IfStatement;

        // Label names.
        let sfx = builder.freshSuffix();
        let thenLab = "then" + sfx;
        let elseLab = "else" + sfx;
        let endLab = "endif" + sfx;

        // Branch.
        let cond = emitExpr(if_.expression);
        builder.buildEffect("br", [cond.dest], undefined, [thenLab, elseLab]);

        // Statement chunks.
        builder.buildLabel(thenLab);
        emit(if_.thenStatement);
        builder.buildEffect("jmp", [], undefined, [endLab]);
        builder.buildLabel(elseLab);
        if (if_.elseStatement) {
          emit(if_.elseStatement);
        }
        builder.buildLabel(endLab);

        break;
      }

      // Plain "for" loops.
      case ts.SyntaxKind.ForStatement: {
        let for_ = node as ts.ForStatement;

        // Label names.
        let sfx = builder.freshSuffix();
        let condLab = "for.cond" + sfx;
        let bodyLab = "for.body" + sfx;
        let endLab = "for.end" + sfx;

        // Initialization.
        if (for_.initializer) {
          emit(for_.initializer);
        }

        // Condition check.
        builder.buildLabel(condLab);
        if (for_.condition) {
          let cond = emitExpr(for_.condition);
          builder.buildEffect("br", [cond.dest], undefined, [bodyLab, endLab]);
        }

        builder.buildLabel(bodyLab);
        emit(for_.statement);
        if (for_.incrementor) {
          emitExpr(for_.incrementor);
        }
        builder.buildEffect("jmp", [], undefined, [condLab]);
        builder.buildLabel(endLab);

        break;
      }

      case ts.SyntaxKind.FunctionDeclaration: 
        let funcDef = node as ts.FunctionDeclaration;
        if (funcDef.name === undefined) {
          throw `no anonymous functions!`;
        }
        let name: string = funcDef.name.getText();
        let args: bril.Argument[] = [];

        for (let p of funcDef.parameters) {
          let argName = p.name.getText();
          let argType = brilType(p, checker);
          args.push({name: argName, type: argType} as bril.Argument);
        }

        // The type checker gives a full function type;
        // we want only the return type.
        if (funcDef.type && funcDef.type.getText() !== 'void') {
          let retType: bril.Type;
          if (funcDef.type.getText() === 'number') {
            retType = "float";
          } else if (funcDef.type.getText() === 'boolean') {
            retType = "bool";
          } else if (funcDef.type.getText() === 'bigint') {
            retType = "int";
          } else {
            throw `unsupported type for function return: ${funcDef.type}`;
          }
          builder.buildFunction(name, args, retType);
        } else {
          builder.buildFunction(name, args);
        }
        if (funcDef.body) {
          emit(funcDef.body);
        }
        break;

      case ts.SyntaxKind.ReturnStatement: {
        let retstmt = node as ts.ReturnStatement;
        if (retstmt.expression) {
          let val = emitExpr(retstmt.expression);
          builder.buildEffect("ret", [val.dest]);
        } else {
          builder.buildEffect("ret", []);
        }
        break;
      }

      default:
        throw `unhandled TypeScript AST node kind ${node.kind}`;
    }
  }

  emit(prog);
  return builder.program;
}

function main() {
  // Get the TypeScript filename.
  let filename = process.argv[2];
  if (!filename) {
    console.error(`usage: ${process.argv[1]} src.ts`)
    process.exit(1);
  }

  // Load up the TypeScript context.
  let program = ts.createProgram([filename], {
    target: ts.ScriptTarget.ES5,
  });
  let checker = program.getTypeChecker();

  // Do a weird dance to look up our source file.
  let sf: ts.SourceFile | undefined;
  for (let file of program.getSourceFiles()) {
    if (file.fileName === filename) {
      sf = file;
      break;
    }
  }
  if (!sf) {
    throw "source file not found";
  }

  // Generate Bril code.
  let brilProg = emitBril(sf, checker);
  process.stdout.write(
    JSON.stringify(brilProg, undefined, 2)
  );
}

// Make unhandled promise rejections terminate.
process.on('unhandledRejection', e => { throw e });

main();
