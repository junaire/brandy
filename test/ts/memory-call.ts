import * as mem from "../../bril-ts/mem";

function update(p: mem.Pointer<bigint>): mem.Pointer<bigint> {
    mem.store(p, 42n);
    return p;
}

let p1 = mem.alloc<bigint>(1n);
let p2 = update(p1);
console.log(mem.load(p2));

mem.free(p1);