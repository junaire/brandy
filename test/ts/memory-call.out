@main {
  v4: int = const 1;
  v5: ptr<int> = alloc v4;
  p1: ptr<int> = id v5;
  v6: ptr<int> = id p1;
  p2: ptr<int> = call @update v6;
  p2: ptr<int> = id p2;
  v7: ptr<int> = id p2;
  v8: int = load v7;
  print v8;
  v9: int = const 0;
  v10: ptr<int> = id p1;
  free v10;
  v11: int = const 0;
}
@update(p: ptr<int>): ptr<int> {
  v0: ptr<int> = id p;
  v1: int = const 42;
  store v0 v1;
  v2: int = const 0;
  v3: ptr<int> = id p;
  ret v3;
}
